accord-core/src/main/java/accord/coordinate/Persist.java

```diff
@@ -21,73 +21,87 @@
+        node.send(sendTo.nodes(), to -> new Apply(to, sendTo, applyTo, executeAt.epoch, txnId, route, executeAt, deps, writes, result), persist);
     }
 
-    public static void persistAndCommit(Node node, TxnId txnId, Key homeKey, Txn txn, Timestamp executeAt, Deps deps, Writes writes, Result result)
+    public static void persistAndCommit(Node node, TxnId txnId, Route route, Txn txn, Timestamp executeAt, Deps deps, Writes writes, Result result)
     {
-        Topologies persistTo = node.topology().preciseEpochs(txn, executeAt.epoch);
-        Persist persist = new Persist(node, persistTo, txnId, homeKey, executeAt);
-        node.send(persistTo.nodes(), to -> new Apply(to, persistTo, txnId, txn, homeKey, executeAt, deps, writes, result), persist);
-        if (txnId.epoch != executeAt.epoch)
-        {
-            Topologies earlierTopologies = node.topology().preciseEpochs(txn, txnId.epoch, executeAt.epoch - 1);
-            Commit.commit(node, earlierTopologies, persistTo, txnId, txn, homeKey, executeAt, deps);
-        }
+        Topologies sendTo = node.topology().preciseEpochs(route, txnId.epoch, executeAt.epoch);
+        Topologies applyTo = node.topology().forEpoch(route, executeAt.epoch);
+        Persist persist = new Persist(node, sendTo, txnId, route, txn, executeAt, deps);
+        node.send(sendTo.nodes(), to -> new Apply(to, sendTo, applyTo, executeAt.epoch, txnId, route, executeAt, deps, writes, result), persist);
     }
 
-    private Persist(Node node, Topologies topologies, TxnId txnId, Key homeKey, Timestamp executeAt)
+    private Persist(Node node, Topologies topologies, TxnId txnId, Route route, Txn txn, Timestamp executeAt, Deps deps)
     {
         this.node = node;
         this.txnId = txnId;
-        this.homeKey = homeKey;
+        this.txn = txn;
+        this.deps = deps;
+        this.route = route;
         this.tracker = new QuorumTracker(topologies);
         this.executeAt = executeAt;
         this.persistedOn = new HashSet<>();
     }
 
     @Override
-    public void onSuccess(Id from, ApplyOk response)
+    public void onSuccess(Id from, ApplyReply reply)
     {
-        persistedOn.add(from);
-        if (tracker.success(from) && !isDone)
+        switch (reply)
         {
-            // TODO: send to non-home replicas also, so they may clear their log more easily?
-            Shard homeShard = node.topology().forEpochIfKnown(homeKey, txnId.epoch);
-            node.send(homeShard, new InformOfPersistence(txnId, homeKey, executeAt, persistedOn));
-            isDone = true;
+            default: throw new IllegalStateException();
+            case Redundant:
+            case Applied:
+                persistedOn.add(from);
+                if (tracker.success(from) && !isDone)
+                {
+                    // TODO: send to non-home replicas also, so they may clear their log more easily?
+                    Shard homeShard = node.topology().forEpochIfKnown(route.homeKey, txnId.epoch);
```

Comment:
Is knowing it is Universal valuable? Does logging durability twice have a cost?

Should we wait and inform Durable/Universal once after waiting until we have either heard back or timed out?

Reply:
I expect that eventually we will, on Cassandra side, aggregate these messages to send batch-wise. But we will at least somehow in the meantime want to know when we *can* delete state from our log; and when there’s no reason not to. Perhaps it would be fine for now though to always defer this until we abandon waiting for further replies though.

Reply:
I don't think you need to deal with it as part of this PR.